Source/Core/Core/HW/WII_IPC.cpp

```diff
@@ -200,13 +219,62 @@ void RegisterMMIO(MMIO::Mapping* mmio, u32 base)
                    // TODO: AVE, SLOT_LED
                  }));
   mmio->Register(base | GPIOB_DIR, MMIO::DirectRead<u32>(&gpio_dir.m_hex),
-                 MMIO::DirectWrite<u32>(&gpio_dir.m_hex));
+                 MMIO::ComplexWrite<u32>([](u32, u32 val) {
+                   gpio_dir.m_hex = (val & gpio_owner.m_hex) | (gpio_dir.m_hex & ~gpio_owner.m_hex);
+                 }));
   mmio->Register(base | GPIOB_IN, MMIO::ComplexRead<u32>([](u32) {
                    Common::Flags<GPIO> gpio_in;
                    gpio_in[GPIO::SLOT_IN] = DVDInterface::IsDiscInside();
                    return gpio_in.m_hex;
                  }),
                  MMIO::Nop<u32>());
+  // Starlet GPIO registers, not normally accessible by PPC (but they can be depending on how
+  // AHBPROT is set up).  We just always allow access, since some homebrew uses them.
+
+  // Note from WiiBrew: When switching owners, copying of the data is not necessary. For example, if
+  // pin 0 has certain configuration in the HW_GPIO registers, and that bit is then set in the
+  // HW_GPIO_OWNER register, those settings will immediately be visible in the HW_GPIOB registers.
+  // There is only one set of data registers, and the HW_GPIO_OWNER register just controls the
+  // access that the HW_GPIOB registers have to that data.
+  // Also: The HW_GPIO registers always have read access to all pins, but any writes (changes) must
+  // go through the HW_GPIOB registers if the corresponding bit is set in the HW_GPIO_OWNER
+  // register.
+  mmio->Register(base | GPIO_OUT, MMIO::DirectRead<u32>(&g_gpio_out.m_hex),
+                 MMIO::ComplexWrite<u32>([](u32, u32 val) {
+                   g_gpio_out.m_hex =
+                       (g_gpio_out.m_hex & gpio_owner.m_hex) | (val & ~gpio_owner.m_hex);
+                   if (g_gpio_out[GPIO::DO_EJECT])
+                   {
+                     INFO_LOG_FMT(WII_IPC, "Ejecting disc due to GPIO write");
+                     DVDInterface::EjectDisc(DVDInterface::EjectCause::Software);
+                   }
+                   // SENSOR_BAR is checked by WiimoteEmu::CameraLogic
+                   // TODO: AVE, SLOT_LED
```

Comment:
(How would handling SLOT_LED even work? :eyes:)

Reply:
Maybe for gaming PCs, making the excessive number of LEDs flash bright blue?  There probably isn't a reasonable way to actually implement it, but it's still worth noting that it isn't implemented yet.